added get_qpy_version to retrieve qpy version

arujjval opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

#14201
### Summary
To get the qpy version, I have added a new function, "get_qpy_version," according to the suggestion given in the issue.

### Comments
I am unsure where to write the tests for this if needed. Is there any guide about writing tests?

arujjval committed:
added get_qpy_version

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=14212) <br/>All committers have signed the CLA.

ElePT reviewed (commented):
Hi @arujjval, thanks for your contribution! When someone develops a new feature we require the following items:

- The feature must be tested, I suggest adding a unit test in the following file: `test/python/qpy/test_circuit_load_from_qpy.py`
- If it's a user-facing feature, it must be documented. For this, you should add it to the API docs together with dump and load: https://github.com/Qiskit/qiskit/blob/main/qiskit/qpy/__init__.py#L100
- Finally, you should add a releasenote following the instructions in the contribution guidelines
If you have any question, let us know.

arujjval committed:
Merge branch 'main' into get-qpy-version

arujjval committed:
added test for get_qpy_version

arujjval commented:
@ElePT Thanks for the help!

I have made changes like you asked.

mtreinish commented on releasenotes/notes/get_qpy_version-75791c063af4e851.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+upgrade_qpy:
```
This is a new feature, upgrade notes are typically used for things that require users to take manual action when upgrade to the new release.

```suggestion
features_qpy:
```

mtreinish commented on test/python/qpy/test_circuit_load_from_qpy.py:
```diff
@@ -66,6 +66,16 @@ def test_invalid_qpy_version(self):
+        qpy_file = io.BytesIO()
+        qpy_file.write(b"QISKIT")
+        qpy_file.write(bytes([QPY_VERSION]))
+        qpy_file.seek(0)
```
It would be a better test to call `qpy.dump()` to generate real payloads with different version numbers (you can do a loop over a range from `QPY_COMPATIBILITY_VERSION` to `QPY_VERSION`) to ensure it's reading a real payload instead of the expected header format.

arujjval replied:
Thanks for review. 

I have made suggested changes.

mtreinish reviewed (commented):
Code-wise the new function looks good, I just had a suggestion inline on the release notes and the test code. Also it looks like CI is failing on either lint or docs (although the logs have expired).

arujjval committed:
Merge branch 'Qiskit:main' into get-qpy-version

arujjval committed:
updated test_qpy_versions with different version numbers

arujjval committed:
resolved lint errors

mtreinish commented on releasenotes/notes/get_qpy_version-75791c063af4e851.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+features_qpy:
+  - |
+    Added a new function `get_qpy_version()` to the `qpy` module. This function retrieves the version of the QPy package currently in use. 
```
```suggestion
    Added a new function :func:`.get_qpy_version` to the :func:`.qpy` module. This function will inspect a QPY file
    and retrieve the QPY format version used in the payload.
```

mtreinish commented on qiskit/qpy/interface.py:
```diff
@@ -342,3 +342,20 @@ def load(
+            data for a circuit.
+
+    Returns:
+        The current QPY version.
```
```suggestion
        The QPY version of the specified file.
```

mtreinish commented on qiskit/qpy/interface.py:
```diff
@@ -342,3 +342,20 @@ def load(
+) -> int:
+    """This function identifies the QPY version of the file.
+
+    Args:
```
We should document that this reads the first 7 bytes of the file and advances the cursor that far. This way users know that if they read this and want to use the file elsewhere (like qpy.load) they'll need to call seek on the file.

arujjval replied:
```
This function identifies the QPY version of the file. It achieves this by reading the
first 7 bytes of the file and unpacking the version number.

Since it advances the cursor to read that, "qpy.load()" should be called after this
function to read the file and use it where needed.
```
  
  Does this look sufficient for documentation?

mtreinish replied:
The second paragraph isn't quite correct. To use `qpy.load` you need to move the cursor back 7 bytes. If you pass it without doing that the load function will error because it doesn't think the file is valid qpy starting 7 bytes in.

mtreinish replied:
I decided we should change this behavior and not advance the cursor as part of this function. I made this change and added a docstring and test for it in: https://github.com/Qiskit/qiskit/pull/14212/commits/1f151d663f6efe52f35af2f1155bb9d5d646682b

mtreinish reviewed (commented):
Thanks for the updates, just a few small comments on how we cn improve the docs a little bit to make the usage a bit more clear.

arujjval committed:
Merge branch 'Qiskit:main' into get-qpy-version

mtreinish committed:
Apply suggestions from code review

mtreinish committed:
Merge branch 'main' into get-qpy-version

mtreinish committed:
Do not advance cursor on exit

This commit reworks the internal logic to ensure we're not advancing the
file cursor on exiting the function. If we didn't seek back the 7 bytes
we read this would be forcing pretty much every conceivable usage of
would be using this as a check and then loading the qpy based on the
result. This behavior is then added to the docstring and a test is added
to assert this behavior is part of the api.

mtreinish reviewed (approved):
This LGTM now, but I think since I touched a fair amount of the code now getting someone else to give it final approval for merge would be prudent.
